added a lm-based optimizer for kornia-pnp

5usu opened the pull request:
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/9fb07eea-fb16-4e93-9c38-b67725be0145" />

- Added LM-based pose refinement (`src/refine.rs`) with `LMParams` and `refine_pose_lm` (numeric Jacobian + 6×6 solver).
- Integrated LM into EPnP via `EPnPParams { refine_lm: Option<LMParams> }`; defaults to off.
- Exported `refine` module and `LMParams` in `src/lib.rs`.
- Updated `EPnP` to optionally run LM and return updated RMSE/iterations/convergence.
- Fixed minor glam types in `refine.rs`.
- Enhanced `examples/pnp_demo` to compare EPnP vs EPnP+LM (printed metrics and red/blue reprojections).

5usu commented:
done revived kornia-py :)

cjpurackal commented:
I'm still seeing a discrepancy of the cube after LM refinement. I'd expect the observed cube (green) and the cube after LM refinement (blue) to fully align.
<img width="2834" height="1468" alt="image" src="https://github.com/user-attachments/assets/4542a030-a564-4833-af46-3c6e222b9daa" />

cjpurackal commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,260 @@
+
+    // Helper closures
+    let project_all = |x: &[f32; 6]| -> (Vec<f32>, f32) {
+        let r = rodrigues(&[x[0], x[1], x[2]]);
```
use `SO3::exp` instead.
eg:
```
SO3::exp(Vec3A::from_array(*rvec)).matrix()
```

cjpurackal commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,260 @@
+/// - `rvec`: Initial axis-angle rotation (input/output)
+/// - `t`: Initial translation (input/output)
+///
+/// Returns updated `(rvec, t, rmse, num_iters, converged)`.
```
Docs mismatch with return

cjpurackal commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,260 @@
+        let t_vec = Vec3::new(x[3], x[4], x[5]);
+        let r_mat = r;
+
+        let mut residuals = Vec::with_capacity(2 * n);
```
Put this outside the closure

cjpurackal commented:
Can you first cross check and verify this with opencv ?

5usu commented:
> Can you first cross check and verify this with opencv ?

cool I'll make the changes and verify with opencv :)

cjpurackal commented:
> - Replaced custom Rodrigues with SO3::exp for rotations; removed redundant temps.
> 
> - Fixed minor issues: removed redundant assignment in epnp.rs, corrected refine_pose_lm docs.
> 
> - Stabilized Jacobian: switched to central differences with tuned step sizes (rot/translation).
> 
> 
> 
> @cjpurackal 

How does it look after the updates ?

5usu commented:
> > * Replaced custom Rodrigues with SO3::exp for rotations; removed redundant temps.
> > * Fixed minor issues: removed redundant assignment in epnp.rs, corrected refine_pose_lm docs.
> > * Stabilized Jacobian: switched to central differences with tuned step sizes (rot/translation).
> > 
> > @cjpurackal
> 
> How does it look after the updates ?

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/e1f78daf-b1cf-46cf-8d46-f47505a83070" />

edgarriba commented on crates/kornia-pnp/src/lib.rs:
```diff
@@ -7,8 +7,12 @@ pub mod epnp;
 pub mod pnp;
 
+/// Optional refinement routines (Levenberg–Marquardt, etc.)
+pub mod refine;
```
are we sure we want this public ?

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,240 @@
+    let mut converged = false;
+
+    while iters < params.max_iters {
+        iters += 1;
```
shouldn't be this in the end of the loop after suceed

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,240 @@
+    while iters < params.max_iters {
+        iters += 1;
+        // Numerical Jacobian J (2N x 6)
+        let mut j = vec![0.0f32; 2 * n * 6];
```
this could be preallocated and just set to zero each iteration

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,240 @@
+        let mut j = vec![0.0f32; 2 * n * 6];
+        let h_rot = 1e-4f32; // radians
+        let t_scale = x[3].abs().max(x[4].abs()).max(x[5].abs()).max(1.0);
+        let h_trans = 1e-4f32 * t_scale; // world units
```
use a const for magic numbers

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,240 @@
+        }
+
+        // Build normal equations: (J^T J + lambda I) delta = -J^T r
+        let mut a = [0.0f32; 36];
```
same for pre allocation

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,240 @@
+// Rodrigues helpers removed; use SO3::exp for rotations.
+
+// Dense 6x6 solver using Gaussian elimination with partial pivoting.
+fn solve_6x6(a: &mut [f32; 36], b: &mut [f32; 6]) -> Option<[f32; 6]> {
```
should we put this later in linalg? @cjpurackal

cjpurackal replied:
Later, yes.

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    // Write back results
+    rvec.copy_from_slice(&[x[0], x[1], x[2]]);
```
rvec.copy_from_slice(&x);

edgarriba commented on examples/pnp_demo/src/main.rs:
```diff
@@ -107,26 +113,35 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
-        img_reproj.push((u, v));
-    }
+    // Helper to reproject with a given pose
+    let reproject = |r: [[f32; 3]; 3], t: [f32; 3]| -> Vec<(f32, f32)> {
```
i advocate to create later a project_points thing similar to opencv, using glam to make use of ismd instructions

5usu commented:
@edgarriba  can you review latest

5usu commented:
> @5usu Can you also add a test for LM refinement ?

done

cjpurackal commented:
cool, please address the other comments as well and i think we're good to go.

5usu commented:
I guess I'm done with all the changes now

5usu commented:
@edgarriba can you review?

edgarriba commented on crates/kornia-imgproc/src/bin/harris_u8.rs:
this is unrelated for this PR

edgarriba commented on crates/kornia-imgproc/src/features/responses.rs:
this is unrelated for this PR

edgarriba commented on crates/kornia-pnp/src/epnp.rs:
```diff
@@ -31,6 +32,9 @@ impl PnPSolver for EPnP {
 pub struct EPnPParams {
     /// Shared numeric tolerances.
     pub tol: NumericTol,
+    /// Optional Levenberg–Marquardt refinement parameters. If `Some`, a final LM pass
```
```suggestion
    /// Optional Levenberg–Marquardt refinement parameters. If given, a final LM pass
```

edgarriba commented on crates/kornia-pnp/src/epnp.rs:
```diff
@@ -136,15 +140,38 @@ pub fn solve_epnp(
+
+    let mut rmse = best_err;
+    let mut iterations = None;
+    let mut converged = Some(true);
```
why converged has to be optional ? @cjpurackal

edgarriba replied:
same for num of iterations

edgarriba commented on crates/kornia-pnp/src/lib.rs:
```diff
@@ -7,8 +7,12 @@ pub mod epnp;
 /// Common data types shared across PnP solvers.
 pub mod pnp;
 
+/// Optional refinement routines (Levenberg–Marquardt, etc.)
```
```suggestion
/// refinement routines (Levenberg–Marquardt, etc.)
```

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,333 @@
+}
+
+/// Refine a pose (rvec, t) with Levenberg–Marquardt to minimize pixel reprojection error.
+///
```
```suggestion
/// # Arguments
```
follow same convention as in other parts of the library

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,333 @@
+    let mut lambda = params.lambda_init;
+    let mut err_sq_base = project_all_in_place(&x, &mut residuals);
+
+    let mut iters = 0usize;
```
```suggestion
    let mut iters = 0;
```
type not recognized by the compiler ?

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,333 @@
+    let mut converged = false;
+
+    // Preallocate J, A and b once
+    let mut j = vec![0.0f32; RESIDUAL_DIM * n * SOLVER_DIM];
```
why not an array in the stack as `a` and `b` ?

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,333 @@
+        let mut rhs = [0.0f32; SOLVER_DIM];
+        for i in 0..SOLVER_DIM {
+            rhs[i] = -b[i];
+        }
```
```suggestion
        }

```

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,333 @@
+                    a[c * SOLVER_DIM + d] += j_ic * j[r_i * SOLVER_DIM + d];
+                }
+            }
+        }
```
```suggestion
        }

```

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,333 @@
+
+        // Solve A delta = -b
+        let mut rhs = [0.0f32; SOLVER_DIM];
+        for i in 0..SOLVER_DIM {
```
why not use iterators with zip, it's more rust idiomatic and prevents random access every time

edgarriba commented on crates/kornia-pnp/src/refine.rs:
```diff
@@ -0,0 +1,333 @@
+            // Tentative update
+            let mut x_new = x;
+            for i in 0..SOLVER_DIM {
+                x_new[i] += delta[i];
```
why not use iterators with zip, it's more rust idiomatic and prevents random access every time

edgarriba commented on examples/pnp_demo/src/main.rs:
```diff
@@ -143,11 +158,21 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
-        println!("Reprojection RMSE: {:.3} px", rmse);
+    println!("Estimated rotation (EPnP):\n{:?}", result_epnp.rotation);
+    println!("Estimated rotation (LM):\n{:?}", result_lm.rotation);
+    if let Some(rmse) = result_epnp.reproj_rmse {
```
@cjpurackal i feel the result of pnp should be Optional. And then inside if lm was requested, this being also optional. Every attribute being optional is not a good pattern

5usu commented:
I have resolved the unrelated PRs @cjpurackal

cjpurackal commented:
@5usu Can you please take a look at Edgar's comments?

edgarriba commented on crates/kornia-pnp/src/epnp.rs:
```diff
@@ -151,15 +155,38 @@ pub fn solve_epnp(
+            ];
+            best_r = r_ref;
+            rmse = rmse_new;
+            iterations = Some(iters);
```
it feels a bit weird in terms of api that iterations and converged are optional. I would upgrade the Result struct packing the this two 

```rust
#[derive(Debug, Clone)]
pub struct RefinementResult {
   pub num_iterations: usize,
   pub converged: boo,
}

#[derive(Debug, Clone)]
pub struct PnPResult {
    /// Estimated rotation matrix.
    pub rotation: [[f32; 3]; 3],
    /// Estimated translation vector.
    pub translation: [f32; 3],
    /// Rodrigues axis-angle representation of the rotation.
    pub rvec: [f32; 3],
    /// Root-mean-square reprojection error in pixels (if computed).
    pub reproj_rmse: Option<f32>,
    /// refinenement result
    pub refinement: Option<RefinementResutlt>
}
```

edgarriba commented:
Please, rebase to main and reopen if still interested in merging this
